draft-thomson-webpush-vapid.md

```diff
@@ -132,211 +134,279 @@ The words "MUST", "MUST NOT", "SHOULD", and "MAY" are used in this document.
+Extra whitespace is added to meet formatting constraints.
 
-## Client-Vended Tokens {#client-token}
+This equates to a JWT with the header and body shown in {{ex-jwt}}.  This JWT
+would be valid until 2016-01-21T01:53:25Z [RFC3339].
 
-In this model, clients generates a token that it uses to prove ownership over a
-private key.  Use of the same key over time establishes a continuous identity.
+~~~
+header = {"typ":"JWT","alg":"ES256"}
+body = { "aud":"https://push.example.net",
+         "exp":1453341205,
+         "iss":"mailto:push@example.com" }
+~~~
+{: #ex-jwt title="Example JWT Header and Body"}
 
-Push message requests can be accompanied by a JSON Web Token (JWT) [RFC7519].
-An "aud" (Audience) claim in the token MUST include the push resource URL.  This
-binds the token to a specific push subscription, but not a specific push
-message.  As a result, the token is reusable, limited by the value of the "exp"
-(Expiry) claim in the token.  An "exp" claim MUST be included.
 
-The JWT is included in an Authorization header field, using an auth-scheme of
-"WebPush".
+# WebPush Authentication Scheme {#auth}
 
-Editor's Note:
+A new "WebPush" HTTP authentication scheme [RFC7235] is defined.  This
+authentication scheme carries a signed JWT, as described in {{jwt}}.
 
-: The definition of the "Bearer" auth-scheme in [RFC6750] is almost perfect, but
-  context would seem to indicate that this is only valid for use with OAuth.
+This authentication scheme is for origin-server authentication only.  Therefore,
+this authentication scheme MUST NOT be used with The Proxy-Authenticate or
+Proxy-Authorization header fields.
 
-The corresponding public key is included in a JSON Web Key (JWK) [RFC7517].
-This would be included in either a newly-defined "jwk" parameter of the
-Crypto-Key header field [I-D.ietf-httpbis-encryption-encoding]; alternatively, the
-"p256ecdsa" parameter defined in [I-D.thomson-http-content-signature] could be
-used to transport a raw key.
+This authentication scheme does not require a challenge.  Clients are able to
+generate the Authorization header field without any additional information from
+a server.  Therefore, a challenge for this authentication scheme MUST NOT be
+sent in a WWW-Authenticate header field.
 
-For voluntarily-provided contact details, a separate header field could be used
-(as in {{from}}) or the JWT could include claims about identity.  For the
-latter, the "sub" (subject) claim could include a contact URI for the
-application server.
+A single parameter is defined for this challenge.  A `keyid` parameter includes
+an identifier for the key that was used to sign the JWT.  This parameter MAY be
+omitted if the recipient can be expected to locate the correct key without this
+assistance.  For instance, if there is exactly one Crypto-Key header field in
+the same request that contains a key of the right type, the `keyid` can be
+safely omitted.
```

Comment:
I honestly can't think of one.  I added this keyid stuff when I was considering having the key omitted from a push request when the push service already knows the key (the "subscription association" case).  But I think that it's just noise now and I will remove it.